libs/langchain/langchain/text_splitter.py

```diff
@@ -130,27 +134,40 @@ def __init__(
         self._keep_separator = keep_separator
         self._add_start_index = add_start_index
         self._strip_whitespace = strip_whitespace
+        self._use_multithreading = use_multithreading
 
     @abstractmethod
     def split_text(self, text: str) -> List[str]:
         """Split text into multiple components."""
 
+    def _create_document(self, args: Tuple[List[str], Optional[List[dict]]]) -> List[Document]:
+        text, metadata = args
+        documents = []
+        index = -1
+        for chunk in self.split_text(text):
+            _metadata = copy.deepcopy(metadata)
+            if self._add_start_index:
+                index = text.find(chunk, index + 1)
+                _metadata["start_index"] = index
+            new_doc = Document(page_content=chunk, metadata=_metadata)
+            documents.append(new_doc)
+        return documents
+    
     def create_documents(
-        self, texts: List[str], metadatas: Optional[List[dict]] = None
+            self, texts: List[str], metadatas: Optional[List[dict]] = None
     ) -> List[Document]:
         """Create documents from a list of texts."""
         _metadatas = metadatas or [{}] * len(texts)
-        documents = []
-        for i, text in enumerate(texts):
-            index = -1
-            for chunk in self.split_text(text):
-                metadata = copy.deepcopy(_metadatas[i])
-                if self._add_start_index:
-                    index = text.find(chunk, index + 1)
-                    metadata["start_index"] = index
-                new_doc = Document(page_content=chunk, metadata=metadata)
-                documents.append(new_doc)
-        return documents
+
+        if self._use_multithreading:
+            with ThreadPoolExecutor() as executor:
```

Comment:
@gustavz Correct me if I'm wrong, but I think that this will end up being slower rather than faster due to the python GIL. 

This will only start working fast with python 3.13 if running with gil disabled: https://peps.python.org/pep-0703/

Speed up can be achieved using a process pool, but we'd probably want to design this carefully as the `create_documents` is operating on a batch rather than a generator and it's unclear to what batch size client code will be tuned. (Though maybe not an issue if the user can toggle the option on/off.)